File: examples/python/prj-output/output_projector_doc.py
"""Output projector examples for documentation.

This file contains simplified examples used in the projector documentation,
demonstrating both OO-style and StateRouter patterns.
"""

from typing import Dict

from angzarr_client import Projector, StateRouter
from angzarr_client.proto.examples import player_pb2 as player
from angzarr_client.proto.examples import hand_pb2 as hand


# docs:start:projector_oo
class OutputProjector(Projector):
    def __init__(self):
        self.player_names: Dict[str, str] = {}

    def handle_player_registered(self, event: player.PlayerRegistered):
        self.player_names[event.player_id] = event.display_name
        print(f"[Player] {event.display_name} registered")

    def handle_funds_deposited(self, event: player.FundsDeposited):
        name = self.player_names.get(event.player_id, event.player_id)
        amount = event.amount.amount if event.amount else 0
        print(f"[Player] {name} deposited ${amount / 100:.2f}")

    def handle_cards_dealt(self, event: hand.CardsDealt):
        for player_cards in event.player_cards:
            name = self.player_names.get(player_cards.player_id, player_cards.player_id)
            cards = format_cards(player_cards.hole_cards)
            print(f"[Hand] {name} dealt {cards}")
# docs:end:projector_oo


def format_cards(cards) -> str:
    return " ".join(f"{c.rank}{c.suit}" for c in cards)


# docs:start:state_router
player_names: Dict[str, str] = {}


def handle_player_registered(event: player.PlayerRegistered):
    player_names[event.player_id] = event.display_name
    print(f"[Player] {event.display_name} registered")


def handle_funds_deposited(event: player.FundsDeposited):
    name = player_names.get(event.player_id, event.player_id)
    print(f"[Player] {name} deposited ${event.amount.amount / 100:.2f}")


def handle_cards_dealt(event: hand.CardsDealt):
    for pc in event.player_cards:
        name = player_names.get(pc.player_id, pc.player_id)
        print(f"[Hand] {name} dealt cards")


router = (
    StateRouter("prj-output")
    .subscribes("player", ["PlayerRegistered", "FundsDeposited"])
    .subscribes("hand", ["CardsDealt", "ActionTaken", "PotAwarded"])
    .on("PlayerRegistered", handle_player_registered)
    .on("FundsDeposited", handle_funds_deposited)
    .on("CardsDealt", handle_cards_dealt)
)
# docs:end:state_router

File: examples/rust/prj-output/src/output_projector_doc.rs
//! Output projector examples for documentation.
//!
//! This file contains simplified examples used in the projector documentation,
//! demonstrating both OO-style and StateRouter patterns.

use std::collections::HashMap;

use angzarr_client::proto::examples::{CardsDealt, FundsDeposited, PlayerRegistered};
use angzarr_client::StateRouter;

// docs:start:projector_oo
pub struct OutputProjector {
    player_names: HashMap<String, String>,
}

impl OutputProjector {
    pub fn handle_player_registered(&mut self, event: &PlayerRegistered) {
        self.player_names.insert(event.player_id.clone(), event.display_name.clone());
        println!("[Player] {} registered", event.display_name);
    }

    pub fn handle_funds_deposited(&mut self, event: &FundsDeposited) {
        let name = self.player_names.get(&event.player_id)
            .map(|s| s.as_str())
            .unwrap_or(&event.player_id);
        let amount = event.amount.as_ref().map(|a| a.amount).unwrap_or(0);
        println!("[Player] {} deposited ${:.2}", name, amount as f64 / 100.0);
    }

    pub fn handle_cards_dealt(&mut self, event: &CardsDealt) {
        for player in &event.player_cards {
            let name = self.player_names.get(&player.player_id)
                .map(|s| s.as_str())
                .unwrap_or(&player.player_id);
            let cards = format_cards(&player.hole_cards);
            println!("[Hand] {} dealt {}", name, cards);
        }
    }
}
// docs:end:projector_oo

fn format_cards(cards: &[angzarr_client::proto::examples::Card]) -> String {
    cards.iter().map(|c| format!("{}{}", c.rank, c.suit)).collect::<Vec<_>>().join(" ")
}

// docs:start:state_router
fn build_router() -> StateRouter {
    StateRouter::new("prj-output")
        .subscribes("player", &["PlayerRegistered", "FundsDeposited"])
        .subscribes("hand", &["CardsDealt", "ActionTaken", "PotAwarded"])
        .on::<PlayerRegistered>(handle_player_registered)
        .on::<FundsDeposited>(handle_funds_deposited)
        .on::<CardsDealt>(handle_cards_dealt)
}

fn handle_player_registered(event: &PlayerRegistered, state: &mut ProjectorState) {
    state.player_names.insert(event.player_id.clone(), event.display_name.clone());
    println!("[Player] {} registered", event.display_name);
}

fn handle_funds_deposited(event: &FundsDeposited, state: &mut ProjectorState) {
    let name = state.player_names.get(&event.player_id)
        .map(|s| s.as_str())
        .unwrap_or(&event.player_id);
    println!("[Player] {} deposited", name);
}

fn handle_cards_dealt(event: &CardsDealt, state: &mut ProjectorState) {
    for player in &event.player_cards {
        let name = state.player_names.get(&player.player_id)
            .map(|s| s.as_str())
            .unwrap_or(&player.player_id);
        println!("[Hand] {} dealt cards", name);
    }
}

struct ProjectorState {
    player_names: HashMap<String, String>,
}
// docs:end:state_router
